app/lib/actions/katello/host/update_system_purpose.rb

```diff
@@ -0,0 +1,31 @@
+module Actions
+  module Katello
+    module Host
+      class UpdateSystemPurpose < Actions::EntryAction
+        def plan(host, service_level, purpose_role, purpose_usage, purpose_addons)
+          fail _("Host %s has not been registered with subscription-manager.") % host.name unless host.subscription_facet
+
+          host.subscription_facet.service_level = service_level unless service_level.nil?
+          host.subscription_facet.purpose_role = purpose_role unless purpose_role.nil?
+          host.subscription_facet.purpose_usage = purpose_usage unless purpose_usage.nil?
+
+          if purpose_addons
+            if purpose_addons.empty?
+              host.subscription_facet.purpose_addons = []
+            else
+              purpose_addon_objects = ::Katello::PurposeAddon.where(name: purpose_addons).first_or_create!
+              host.subscription_facet.purpose_addons << purpose_addon_objects unless purpose_addon_objects.id == host.subscription_facet.purpose_addons.pluck(:id).first
```

Comment:
I think you're trying to prevent setting the purpose addons if they match what was given. I think active record might actually be smart enough to do that already in that it won't issue a query to update them. You can test that quickly in the console and SQL logging will help: `ActiveRecord::Base.logger = Logger.new(STDOUT)
`

Reply:
@jturel 

I tried what you suggested and see the SQL, I am still getting duplicates:

```ruby
[16] pry(main)> foo.subscription_facet.purpose_addons
=> [#<Katello::PurposeAddon:0x000000000d0f2870 id: 1, name: "test">, #<Katello::PurposeAddon:0x000000000c544758 id: 1, name: "test">, #<Katello::PurposeAddon:0x000000000c544758 id: 1, name: "test">]
[17] pry(main)> 
```

Reply:
Interesting! Let's pause on that for now. Have you tested adding multiple addons via the API?

Reply:
Yep I think your code needs to be something more like this `addons = ['one', 'two'].map { |x| Katello::PurposeAddon.find_or_create_by(name: x) }`

Reply:
I was thinking more about your optimization, and what the value of it is. In most cases, if you're using this API, there is going to be a difference in the addons  and we would run queries to confirm that only to update the DB anyway. Since we're only deleting and creating at most a handful of rows, maybe better to do without, which simplifies this code. `subscription_facet.purpose_addons = purpose_addon_objects`. What do you think?

Reply:
Can this work for multiples? I wrote a unit test and it passes:

```ruby
if purpose_addons
            if purpose_addons.empty?
              host.subscription_facet.purpose_addons = []
            else
              purpose_addons.each do |addon|
                purpose_addon_objects = ::Katello::PurposeAddon.where(name: addon).first_or_create!
              binding.pry
                host.subscription_facet.purpose_addons << purpose_addon_objects #unless purpose_addon_objects.id == host.subscription_facet.purpose_addons.pluck(:id).first
              binding.pry
              end
            end
          end
```

Test:

```ruby
def test_update_system_purpose_multiple_addons
      plan_action @action, @host, @service_level, @purpose_role, @purpose_usage, ['Addon One', 'Addon Two']

      assert_equal(@host.subscription_facet.service_level, @service_level)
      assert_equal(@host.subscription_facet.purpose_role, @purpose_role)
      assert_equal(@host.subscription_facet.purpose_usage, @purpose_usage)
      assert_equal(@host.subscription_facet.purpose_addons.count, 2)
      assert_equal(@host.subscription_facet.purpose_addons.pluck(:name).first, 'Addon One')
      assert_equal(@host.subscription_facet.purpose_addons.pluck(:name).second, 'Addon Two')
    end
```

Reply:
Still looking at the duplicate issue

Reply:
For the duplicate issue, can we keep the unless in there since activerecord is not picking up on what you suggested.

Reply:
oh nm I see you using `find_or_create`

Reply:
When we do 'subscription_facet.purpose_addons = purpose_addon_objects' that does a complete replacement - so no duplicates unless those were passed into the api. so you'd want to find a way to make the array elements unique

Reply:
Ah so the `<<` is adding to it and adding the dupes compared to the `=` where it just overrides.

Reply:
> I was thinking more about your optimization, and what the value of it is. In most cases, if you're using this API, there is going to be a difference in the addons and we would run queries to confirm that only to update the DB anyway. Since we're only deleting and creating at most a handful of rows, maybe better to do without, which simplifies this code. `subscription_facet.purpose_addons = purpose_addon_objects`. What do you think?

I am fine with changing it to that, I will work on updating it and the unit tests

Reply:
@jturel updated with the new changes, we got it down to just 2 lines and are creating things that are passed in. Let me know what needs to be changed next.